### PR TITLE
add(google-auth)

### PR DESCRIPTION
**Pull request by slava89**
looks like works for local database, user object saved in mongodb successfully

### CODE REVIEW

**Comment by slava89** [comment]
@andrewdacenko can you explane in few words again, what steps we need to complete now to make auth work also on heroku, not just on local.

**Comment by slava89** [comment]
for some reason it saves the user, but regirects not on profile page, but like failure redirect:
app.get('/auth/google/callback',
  passport.authenticate('google', { failureRedirect: '/' }),
  function(req, res) {
    res.redirect('/profile');
  });

File: server.js [hunk 1]
```diff
@@ -12,4 +12,124 @@ app.use(express.static(__dirname + '/client'));
...
+
+var GoogleStrategy = require('passport-google-oauth2').Strategy;
+
+passport.use(new GoogleStrategy({
+  // clientID: config.google.clientID,
+  // clientSecret: config.google.clientSecret,
+  // callbackURL: config.google.callbackURL
+  clientID: process.env.GOOGLE_CLIENT_ID,
+  clientSecret: process.env.GOOGLE_CLIENT_SECRET,
+  callbackURL: "http://localhost:3000/auth/google/callback"
```
Comment (andrewdacenko) [inline server.js]: @slava89 this is local url. You should use one that is abstract. For heroku it should be like 'https://befriends.herokuapp.com/auth/google/callback'

**Comment by slava89** [comment]
also need a remove database instead on local, need to configure it..

**Comment by slava89** [comment]
@andrewdacenko need your help to check why is auth not working. Looks like its not saving user in section or smth. It gets an object from google account and saves it to the database. But it is not authenticate, sine redirecting with failure result: 

passport.authenticate('google', { failureRedirect: '/' }),
function(req, res) {
res.redirect('/profile');
}); // redirects to "/"

**Comment by slava89** [comment]
Maybe its because i forgot to use session before initializing passport and session?
app.use(session({secret: 'fekfneopgwepge'}));

**Comment by andrewdacenko** [comment]
@slava89 Take a look at this example https://github.com/mstade/passport-google-oauth2/blob/master/example/app.js also be aware of order. And read carefully comments.

**Comment by slava89** [comment]
what comments?

**Comment by JackLip** [comment]
in the code, like this

// Passport session setup.
//   To support persistent login sessions, Passport needs to be able to
//   serialize users into and deserialize users out of the session.  Typically,
//   this will be as simple as storing the user ID when serializing, and finding
//   the user by ID when deserializing.  However, since this example does not
//   have a database of user records, the complete Google profile is
//   serialized and deserialized.

**Comment by andrewdacenko** [comment]
@JackLip Yep

**Comment by slava89** [comment]
When i will back to PC in the evening, i will try to add app.use(session({secret: 'fekfneopgwepge'})); and change the order of actions according to your link, if in will not work, i think your help we be needed.

Can you also give us new issues?

**Comment by andrewdacenko** [comment]
@slava89 Also serialize and deserialize properly

**Comment by slava89** [comment]
Can you explane more specific? What i need to add?

**Comment by andrewdacenko** [comment]
@slava89 Per serialization you need to use profile.id of google account and per deserialization use User model to find user by this id

**Comment by slava89** [comment]
ok, will try to figure out how to do it. But i didn't see in any oath2 examples anything else exept: 
passport.serializeUser(function(user, done) {
  done(null, user);
});
passport.deserializeUser(function(obj, done) {
  done(null, obj);
});

File: server.js [hunk 2]
```diff
@@ -10,6 +12,138 @@ app.use(express.static(__dirname + '/client'));
...
+    });
+
+// config
+
+var GoogleStrategy = require('passport-google-oauth2').Strategy;
+
+passport.use(new GoogleStrategy({
+  clientID: process.env.GOOGLE_CLIENT_ID,
+  clientSecret: process.env.GOOGLE_CLIENT_SECRET,
+  callbackURL: "http://localhost:3000/auth/google/callback"
```
Comment (andrewdacenko) [inline server.js]: Change this to `process.env.GOOGLE_AUTH_CALLBACK_URL` 
and fallback to `"http://localhost:" + port + "/auth/google/callback"`

**Pull request merged**